File: lang/julia/OraBench.jl
#= 
OraBench:
- Author: Konnexions GmbH
- Date: 2021-06-26
=#

module OraBench

using Pkg

Pkg.add("CSV")
Pkg.add("DataFrames")
Pkg.add("Formatting")
Pkg.add("Oracle")
Pkg.add("TimesDates")

using CSV
using DataFrames
using Dates
using Formatting
using Oracle
using TimesDates
using TOML

# ----------------------------------------------------------------------------------
# Definition of the global variables.
# ----------------------------------------------------------------------------------

global BENCHMARK_BATCH_SIZE = 0
global BENCHMARK_COMMENT = ""
global BENCHMARK_CORE_MULTIPLIER = 0
global BENCHMARK_DATABASE = ""
global BENCHMARK_DRIVER = ""
global BENCHMARK_HOST_NAME = ""
global BENCHMARK_ID = ""
global BENCHMARK_LANGUAGE = ""
global BENCHMARK_NUMBER_CORES = 0
global BENCHMARK_NUMBER_PARTITIONS = 0
global BENCHMARK_OS = ""
global BENCHMARK_RELEASE = ""
global BENCHMARK_TRANSACTION_SIZE = 0
global BENCHMARK_TRIALS = 0
global BENCHMARK_USER_NAME = ""
global BULK_DATA = nothing
global BULK_DATA_PARTITIONS = Dict()

global CONNECTION_FETCH_SIZE = 0
global CONNECTION_HOST = ""
global CONNECTION_PASSWORD = ""
global CONNECTION_PORT = 0
global CONNECTION_SERVICE = ""
global CONNECTION_USER = ""
global CONNECTIONS = Dict()

global FILE_BULK_DELIMITER = ""
global FILE_BULK_LENGTH = 0
global FILE_BULK_NAME = ""
global FILE_BULK_SIZE = 0
global FILE_CONFIGURATION_NAME_TOML = "priv/properties/ora_bench_toml.properties"
global FILE_RESULT_DELIMITER = ""
global FILE_RESULT_NAME = ""

global IX_DURATION_INSERT_SUM = 4
global IX_DURATION_SELECT_SUM = 5
global IX_LAST_BENCHMARK = 1
global IX_LAST_QUERY = 3
global IX_LAST_TRIAL = 2

global MEASUREMENT_DATA = Array(["", "", "", 0, 0])

global RESULT_FILE = nothing

global SQL_CREATE = ""
global SQL_DROP = ""
global SQL_INSERT = ""
global SQL_SELECT = ""

# ----------------------------------------------------------------------------------
# Creating the database connections.
# ----------------------------------------------------------------------------------

function create_connections()
    function_name = string(StackTraces.stacktrace()[1].func)
    @debug "Start $(function_name)"

    connection_string =
        CONNECTION_HOST * ":" * string(CONNECTION_PORT) * "/" * CONNECTION_SERVICE

    for partition_key = 1:BENCHMARK_NUMBER_PARTITIONS
        try
            CONNECTIONS[partition_key] =
                Oracle.Connection(CONNECTION_USER, CONNECTION_PASSWORD, connection_string)
            @info "wwe connection " * string(partition_key) * " open"
        catch reason
            @info "partition_key      =" * string(partition_key)
            @info "connection_user    =" * CONNECTION_USER
            @info "connection_password=" * CONNECTION_PASSWORD
            @info "connection_string  =" * connection_string
            error(
                "fatal error: program abort =====> database connect error: '$(string(reason))' <=====",
            )
        end
    end

    @debug "End   $(function_name)"
    nothing
end

# ----------------------------------------------------------------------------------
# Writing the results.
# ----------------------------------------------------------------------------------

function create_result(action, trial_number, sql_statement, start_date_time, sql_operation)
    function_name = string(StackTraces.stacktrace()[1].func)
    @debug "Start $(function_name)"

    end_date_time = TimeDate(now())

    duration_ns = (end_date_time - start_date_time).value

    if sql_operation == "insert"
        MEASUREMENT_DATA[IX_DURATION_INSERT_SUM] += duration_ns
    elseif sql_operation == "select"
        MEASUREMENT_DATA[IX_DURATION_SELECT_SUM] += duration_ns
    end

    write(
        RESULT_FILE,
        BENCHMARK_RELEASE *
        FILE_RESULT_DELIMITER *
        BENCHMARK_ID *
        FILE_RESULT_DELIMITER *
        BENCHMARK_COMMENT *
        FILE_RESULT_DELIMITER *
        BENCHMARK_HOST_NAME *
        FILE_RESULT_DELIMITER *
        string(BENCHMARK_NUMBER_CORES) *
        FILE_RESULT_DELIMITER *
        BENCHMARK_OS *
        FILE_RESULT_DELIMITER *
        BENCHMARK_USER_NAME *
        FILE_RESULT_DELIMITER *
        BENCHMARK_DATABASE *
        FILE_RESULT_DELIMITER *
        BENCHMARK_LANGUAGE *
        FILE_RESULT_DELIMITER *
        BENCHMARK_DRIVER *
        FILE_RESULT_DELIMITER *
        string(trial_number) *
        FILE_RESULT_DELIMITER *
        sql_statement *
        FILE_RESULT_DELIMITER *
        string(BENCHMARK_CORE_MULTIPLIER) *
        FILE_RESULT_DELIMITER *
        string(CONNECTION_FETCH_SIZE) *
        FILE_RESULT_DELIMITER *
        string(BENCHMARK_TRANSACTION_SIZE) *
        FILE_RESULT_DELIMITER *
        string(FILE_BULK_LENGTH) *
        FILE_RESULT_DELIMITER *
        string(FILE_BULK_SIZE) *
        FILE_RESULT_DELIMITER *
        string(BENCHMARK_BATCH_SIZE) *
        FILE_RESULT_DELIMITER *
        action *
        FILE_RESULT_DELIMITER *
        string(start_date_time) *
        FILE_RESULT_DELIMITER *
        string(end_date_time) *
        FILE_RESULT_DELIMITER *
        string(round((end_date_time - start_date_time).value * 0.001)) *
        FILE_RESULT_DELIMITER *
        string(duration_ns) *
        FILE_RESULT_DELIMITER *
        "\n",
    )

    @debug "End   $(function_name)"
    nothing
end

# ----------------------------------------------------------------------------------
# Recording the results of the benchmark - end processing.
# ----------------------------------------------------------------------------------

function create_result_measuring_point_end(
    action,
    trial_number = 0,
    sql_statement = "",
    sql_operation = "",
)
    function_name = string(StackTraces.stacktrace()[1].func)
    @debug "Start $(function_name)"

    if action == "query"
        create_result(
            action,
            trial_number,
            sql_statement,
            MEASUREMENT_DATA[IX_LAST_QUERY],
            sql_operation,
        )
    elseif action == "trial"
        create_result(
            action,
            trial_number,
            sql_statement,
            MEASUREMENT_DATA[IX_LAST_TRIAL],
            sql_operation,
        )
    elseif action == "benchmark"
        create_result(
            action,
            trial_number,
            sql_statement,
            MEASUREMENT_DATA[IX_LAST_BENCHMARK],
            sql_operation,
        )
        close(RESULT_FILE)
    else
        error(
            "fatal error: program abort =====> unknown action='$(action)' status='end' <=====",
        )
    end

    @debug "End   $(function_name)"
    nothing
end

# ----------------------------------------------------------------------------------
# Recording the results of the benchmark - start processing.
# ----------------------------------------------------------------------------------

function create_result_measuring_point_start(action)
    function_name = string(StackTraces.stacktrace()[1].func)
    @debug "Start $(function_name)"

    if action == "query"
        MEASUREMENT_DATA[IX_LAST_QUERY] = TimeDate(now())
    elseif action == "trial"
        MEASUREMENT_DATA[IX_LAST_TRIAL] = TimeDate(now())
    else
        error(
            "fatal error: program abort =====> unknown action='$(action)' status='start' <=====",
        )
    end

    @debug "End   $(function_name)"
    nothing
end

function create_result_measuring_point_start_benchmark()
    function_name = string(StackTraces.stacktrace()[1].func)
    @debug "Start $(function_name)"

    MEASUREMENT_DATA[IX_LAST_BENCHMARK] = TimeDate(now())

    if !(isfile(FILE_RESULT_NAME))
        error(
            "fatal error: program abort =====> result file '$(FILE_RESULT_NAME)' is missing <=====",
        )
    end

    global RESULT_FILE = open(FILE_RESULT_NAME, "a")

    @debug "End   $(function_name)"
    nothing
end

# ----------------------------------------------------------------------------------
# Loading the bulk file into memory.
# ----------------------------------------------------------------------------------

function get_bulk_data_partitions()
    function_name = string(StackTraces.stacktrace()[1].func)
    @debug "Start $(function_name)"

    if FILE_BULK_SIZE < BENCHMARK_NUMBER_PARTITIONS
        error(
            "fatal error: program abort =====> size of the bulk file ($(FILE_BULK_SIZE)) is smaller than the number of partitions ($(BENCHMARK_NUMBER_PARTITIONS)) <=====",
        )
    end

    global BULK_DATA =
        DataFrame(CSV.File(FILE_BULK_NAME, header = 1, delim = FILE_BULK_DELIMITER))

    partition_size = div(FILE_BULK_SIZE, BENCHMARK_NUMBER_PARTITIONS)

    # ----------------------------------------------------------------------------------
    # Loading the bulk file into memory.
    # ----------------------------------------------------------------------------------

    @info "Start Distribution of the data in the partitions"

    last_partition_upper = 0

    for partition_key = 1:BENCHMARK_NUMBER_PARTITIONS
        current_partition_upper = last_partition_upper + partition_size
        if current_partition_upper > FILE_BULK_SIZE
            current_partition_upper = FILE_BULK_SIZE
        end
        BULK_DATA_PARTITIONS[partition_key] =
            last_partition_upper + 1, current_partition_upper
        @info format(
            "Partition p{1:0>5d} contains {2:n} rows",
            partition_key,
            current_partition_upper - last_partition_upper,
        )
        last_partition_upper = current_partition_upper
    end

    @info "End   Distribution of the data in the partitions"

    @debug "End   $(function_name)"
    nothing
end

# ----------------------------------------------------------------------------------
# Loading the configuration parameters into memory.
# ----------------------------------------------------------------------------------

function get_config()
    function_name = string(StackTraces.stacktrace()[1].func)
    @debug "Start $(function_name)"

    config_parser::Dict = TOML.parsefile(FILE_CONFIGURATION_NAME_TOML)

    global BENCHMARK_BATCH_SIZE =
        parse(Int64, config_parser["DEFAULT"]["benchmark_batch_size"])
    global BENCHMARK_COMMENT = config_parser["DEFAULT"]["benchmark_comment"]
    global BENCHMARK_CORE_MULTIPLIER =
        parse(Int64, config_parser["DEFAULT"]["benchmark_core_multiplier"])
    global BENCHMARK_DATABASE = config_parser["DEFAULT"]["benchmark_database"]
    global BENCHMARK_HOST_NAME = config_parser["DEFAULT"]["benchmark_host_name"]
    global BENCHMARK_ID = config_parser["DEFAULT"]["benchmark_id"]
    global BENCHMARK_NUMBER_CORES =
        parse(Int64, config_parser["DEFAULT"]["benchmark_number_cores"])
    global BENCHMARK_NUMBER_PARTITIONS = 1
    # wwe   parse(Int64, config_parser["DEFAULT"]["benchmark_number_partitions"])
    global BENCHMARK_OS = config_parser["DEFAULT"]["benchmark_os"]
    global BENCHMARK_RELEASE = config_parser["DEFAULT"]["benchmark_release"]
    global BENCHMARK_TRANSACTION_SIZE =
        parse(Int64, config_parser["DEFAULT"]["benchmark_transaction_size"])
    global BENCHMARK_TRIALS = parse(Int64, config_parser["DEFAULT"]["benchmark_trials"])
    global BENCHMARK_USER_NAME = config_parser["DEFAULT"]["benchmark_user_name"]

    global CONNECTION_FETCH_SIZE =
        parse(Int64, config_parser["DEFAULT"]["connection_fetch_size"])
    global CONNECTION_HOST = config_parser["DEFAULT"]["connection_host"]
    global CONNECTION_PASSWORD = config_parser["DEFAULT"]["connection_password"]
    global CONNECTION_PORT = parse(Int64, config_parser["DEFAULT"]["connection_port"])
    global CONNECTION_SERVICE = config_parser["DEFAULT"]["connection_service"]
    global CONNECTION_USER = config_parser["DEFAULT"]["connection_user"]

    global FILE_BULK_DELIMITER =
        replace(config_parser["DEFAULT"]["file_bulk_delimiter"], "TAB" => "\t")
    global FILE_BULK_LENGTH = parse(Int64, config_parser["DEFAULT"]["file_bulk_length"])
    global FILE_BULK_NAME = config_parser["DEFAULT"]["file_bulk_name"]
    global FILE_BULK_SIZE = parse(Int64, config_parser["DEFAULT"]["file_bulk_size"])
    global FILE_RESULT_DELIMITER =
        replace(config_parser["DEFAULT"]["file_result_delimiter"], "TAB" => "\t")
    global FILE_RESULT_NAME = config_parser["DEFAULT"]["file_result_name"]

    global SQL_CREATE = config_parser["DEFAULT"]["sql_create"]
    global SQL_DROP = config_parser["DEFAULT"]["sql_drop"]
    global SQL_INSERT = replace(
        replace(config_parser["DEFAULT"]["sql_insert"], ":key" => ":1"),
        ":data" => ":2",
    )
    global SQL_SELECT = config_parser["DEFAULT"]["sql_select"]

    @debug "End   $(function_name)"
    nothing
end

# ----------------------------------------------------------------------------------
# Performing the insert operations.
# ----------------------------------------------------------------------------------

function insert(connection, bulk_data_partition)
    function_name = string(StackTraces.stacktrace()[1].func)
    @debug "Start $(function_name)"

    count = 0
    batch_data = list()

#     for key_data_tuple in bulk_data_partition
#         count += 1
# 
#         if BENCHMARK_BATCH_SIZE == 0:
#             cursor.execute(SQL_SELECT, [key_data_tuple[0], key_data_tuple[1]])
#         else:
#             batch_data.append(key_data_tuple)
#             if count % BENCHMARK_BATCH_SIZE == 0:
#                 cursor.executemany(SQL_SELECT, batch_data)
#                 batch_data = list()
#             end
#         end
# 
#         if BENCHMARK_TRANSACTION_SIZE > 0 and count % BENCHMARK_TRANSACTION_SIZE == 0:
#             connection.commit()
#     end
# 
#     if BENCHMARK_BATCH_SIZE > 0 and batch_data.__len__() > 0:
#         cursor.executemany(SQL_SELECT, batch_data)
#     end
# 
#     if BENCHMARK_TRANSACTION_SIZE == 0 or count % BENCHMARK_TRANSACTION_SIZE != 0:
#         connection.commit()
#     end

    @debug "End   $(function_name)"
    nothing
end

# ----------------------------------------------------------------------------------
# Main function.
# ----------------------------------------------------------------------------------

function main()
    function_name = string(StackTraces.stacktrace()[1].func)
    @debug "Start $(function_name)"

    @info "Start OraBench.jl"

    m = Pkg.Operations.Context().env.manifest
    v = m[findfirst(v -> v.name == "Oracle", m)].version

    global BENCHMARK_DRIVER = "Oracle.jl " * v
    @info "wwe BENCHMARK_DRIVER=" + BENCHMARK_DRIVER
    global BENCHMARK_LANGUAGE = "Julia " * string(Base.VERSION)
    @info "wwe BENCHMARK_LANGUAGE=" + BENCHMARK_LANGUAGE

    # run_benchmark()

    @info "End   OraBench.jl"

    @debug "End   $(function_name)"
    nothing
end

# ----------------------------------------------------------------------------------
# Performing the benchmark run.
# ----------------------------------------------------------------------------------

function run_benchmark()
    function_name = string(StackTraces.stacktrace()[1].func)
    @debug "Start $(function_name)"

    create_result_measuring_point_start_benchmark()

    get_config()

    get_bulk_data_partitions()

    create_connections()

    for trial_number = 1:BENCHMARK_TRIALS
        run_trial(trial_number)
    end

    for partition_key = 1:BENCHMARK_NUMBER_PARTITIONS
        Oracle.close(CONNECTIONS[partition_key])
        @info "wwe connection " * string(partition_key) * " closed"
    end

    create_result_measuring_point_end("benchmark")

    @debug "End   $(function_name)"
    nothing
end

# ----------------------------------------------------------------------------------
# Performing the insert operations.
# ----------------------------------------------------------------------------------

function run_insert(trial_number)
    function_name = string(StackTraces.stacktrace()[1].func)
    @debug "Start $(function_name)"

#     create_result_measuring_point_start('query')
# 
#     threads = list()
# 
#     for partition_key in 1: BENCHMARK_NUMBER_PARTITIONS
#         if BENCHMARK_CORE_MULTIPLIER == 0:
#             insert(connections[partition_key], bulk_data_partitions[partition_key])
#         else
#             thread = threading.Thread(target=insert, args=(config, connections[partition_key], cursors[partition_key], bulk_data_partitions[partition_key],))
#             threads.append(thread)
#             thread.start()
#          end
#     end        
# 
#     if BENCHMARK_CORE_MULTIPLIER > 0
#         for thread in threads:
#             thread.join()
#         end
#     end        
# 
#     create_result_measuring_point_end('query', trial_number, SQL_SELECT, 'insert')

    @debug "End   $(function_name)"
    nothing
end

# ----------------------------------------------------------------------------------
# Performing one trial.
# ----------------------------------------------------------------------------------

function run_trial(trial_number)
    function_name = string(StackTraces.stacktrace()[1].func)
    @debug "Start $(function_name)"

    create_result_measuring_point_start("trial")

    @info "Start trial no. $(string(trial_number))"

    try
        Oracle.execute(CONNECTIONS[1], SQL_CREATE)
        @debug "last DDL statement=$(SQL_CREATE)"
    catch
        Oracle.execute(CONNECTIONS[1], SQL_DROP)
        Oracle.execute(CONNECTIONS[1], SQL_CREATE)
        @debug "last DDL statement after DROP=$(SQL_CREATE)"
    end

            run_insert(
                trial_number,
            )

    #     run_select(trial_number)

    Oracle.execute(CONNECTIONS[1], SQL_DROP)
    @debug "last DDL statement=$(SQL_DROP)"

    create_result_measuring_point_end("trial", trial_number)

    @debug "End   $(function_name)"
    nothing
end


# ----------------------------------------------------------------------------------
# Entry point.
# ----------------------------------------------------------------------------------

if abspath(PROGRAM_FILE) == @__FILE__
    main()
end

end

File: src_python/OraBench.py
import configparser
import csv
import cx_Oracle
import datetime
import locale
import logging
import logging.config
import os
import sys
import threading
import yaml
from pathlib import Path

# ------------------------------------------------------------------------------
# Definition of the global variables.
# ------------------------------------------------------------------------------

BENCHMARK_DRIVER = 'Oracle cx_Oracle (Version v' + cx_Oracle.version + ')'
BENCHMARK_LANGUAGE = 'Python 3 ' + sys.version

FILE_CONFIGURATION_NAME_PYTHON = 'priv/properties/ora_bench_python.properties'

IX_DURATION_INSERT_SUM = 3
IX_DURATION_SELECT_SUM = 4
IX_LAST_BENCHMARK = 0
IX_LAST_QUERY = 2
IX_LAST_TRIAL = 1


# ------------------------------------------------------------------------------
# Creating the database objects connection and cursor.
# ------------------------------------------------------------------------------

def create_database_objects(logger, config):
    if logger.isEnabledFor(logging.DEBUG):
        logger.debug('Start')

    connections = list()
    cursors = list()

    for _ in range(0, config['benchmark.number.partitions']):
        try:
            connection = cx_Oracle.connect(config['connection.user'], config['connection.password'],
                                           config['connection.host'] + ':' + str(config['connection.port']) + '/' + config['connection.service'])
            connection.autocommit = False
        except cx_Oracle.DatabaseError as reason:
            logger.info('connection.host    =' + config['connection.host'])
            logger.info('connection.port    =' + str(config['connection.port']))
            logger.info('connection.service =' + config['connection.service'])
            logger.info('connection.user    =' + config['connection.user'])
            logger.info('connection.password=' + config['connection.password'])
            sys.exit('database connect error: ' + str(reason))

        connections.append(connection)
        cursors.append(connection.cursor())

    connections_cursors = (connections, cursors)

    if logger.isEnabledFor(logging.DEBUG):
        logger.debug('End')

    return connections_cursors


# ------------------------------------------------------------------------------
# Writing the results.
# ------------------------------------------------------------------------------

def create_result(logger, config, result_file, measurement_data, action, trial_number, sql_statement, start_date_time, sql_operation):
    if logger.isEnabledFor(logging.DEBUG):
        logger.debug('Start')

    global IX_DURATION_INSERT_SUM
    global IX_DURATION_SELECT_SUM

    end_date_time = datetime.datetime.now()

    duration_ns = (end_date_time - start_date_time).total_seconds() * 1000000000

    if sql_operation == 'insert':
        measurement_data[IX_DURATION_INSERT_SUM] += duration_ns
    elif sql_operation == 'select':
        measurement_data[IX_DURATION_SELECT_SUM] += duration_ns

    result_file.write(config['benchmark.release'] + config['file.result.delimiter'] +
                      config['benchmark.id'] + config['file.result.delimiter'] +
                      config['benchmark.comment'] + config['file.result.delimiter'] +
                      config['benchmark.host.name'] + config['file.result.delimiter'] +
                      str(config['benchmark.number.cores']) + config['file.result.delimiter'] +
                      config['benchmark.os'] + config['file.result.delimiter'] +
                      config['benchmark.user.name'] + config['file.result.delimiter'] +
                      config['benchmark.database'] + config['file.result.delimiter'] +
                      BENCHMARK_LANGUAGE + config['file.result.delimiter'] +
                      BENCHMARK_DRIVER + config['file.result.delimiter'] +
                      str(trial_number) + config['file.result.delimiter'] +
                      sql_statement + config['file.result.delimiter'] +
                      str(config['benchmark.core.multiplier']) + config['file.result.delimiter'] +
                      str(config['connection.fetch.size']) + config['file.result.delimiter'] +
                      str(config['benchmark.transaction.size']) + config['file.result.delimiter'] +
                      str(config['file.bulk.length']) + config['file.result.delimiter'] +
                      str(config['file.bulk.size']) + config['file.result.delimiter'] +
                      str(config['benchmark.batch.size']) + config['file.result.delimiter'] +
                      action + config['file.result.delimiter'] +
                      start_date_time.strftime('%Y-%m-%d %H:%M:%S.%f000') + config['file.result.delimiter'] +
                      end_date_time.strftime('%Y-%m-%d %H:%M:%S.%f000') + config['file.result.delimiter'] +
                      str(round((end_date_time - start_date_time).total_seconds())) + config['file.result.delimiter'] +
                      str(round(duration_ns)) + '\n')

    if logger.isEnabledFor(logging.DEBUG):
        logger.debug('End')


# ------------------------------------------------------------------------------
# Creating the result file.
# ------------------------------------------------------------------------------

def create_result_file(logger, config):
    if logger.isEnabledFor(logging.DEBUG):
        logger.debug('Start')

    result_file = Path(config['file.result.name'])

    if not result_file.is_file():
        logger.error('fatal error: program abort =====> result file "' + config['file.result.name'] + '" is missing <=====')
        sys.exit(1)

    result_file = open(os.path.abspath(config['file.result.name']), 'a')

    if logger.isEnabledFor(logging.DEBUG):
        logger.debug('End')

    return result_file


# ------------------------------------------------------------------------------
# Recording the results of the benchmark - end processing.
# ------------------------------------------------------------------------------

def create_result_measuring_point_end(logger, config, result_file, measurement_data, action, trial_number=0, sql_statement='', sql_operation=''):
    if logger.isEnabledFor(logging.DEBUG):
        logger.debug('Start')

    global IX_LAST_BENCHMARK
    global IX_LAST_QUERY
    global IX_LAST_TRIAL

    if action == 'query':
        create_result(logger, config, result_file, measurement_data, action, trial_number, sql_statement, measurement_data[IX_LAST_QUERY], sql_operation)
    elif action == 'trial':
        create_result(logger, config, result_file, measurement_data, action, trial_number, sql_statement, measurement_data[IX_LAST_TRIAL], sql_operation)
    elif action == 'benchmark':
        create_result(logger, config, result_file, measurement_data, action, trial_number, sql_statement, measurement_data[IX_LAST_BENCHMARK], sql_operation)
        result_file.close()
    else:
        logger.error('action="' + action + '"' + ' state="end"')

        if logger.isEnabledFor(logging.DEBUG):
            logger.debug('End')

        sys.exit(1)

    if logger.isEnabledFor(logging.DEBUG):
        logger.debug('End')


# ------------------------------------------------------------------------------
# Recording the results of the benchmark - start processing.
# ------------------------------------------------------------------------------

def create_result_measuring_point_start(logger, measurement_data, action):
    if logger.isEnabledFor(logging.DEBUG):
        logger.debug('Start')

    global IX_LAST_QUERY
    global IX_LAST_TRIAL

    if action == 'query':
        measurement_data[IX_LAST_QUERY] = datetime.datetime.now()
    elif action == 'trial':
        measurement_data[IX_LAST_TRIAL] = datetime.datetime.now()
    else:
        logger.error('Unknown action="' + action + '"' + ' state="start"')
        sys.exit(1)

    if logger.isEnabledFor(logging.DEBUG):
        logger.debug('End')

    return measurement_data


def create_result_measuring_point_start_benchmark(logger, config):
    if logger.isEnabledFor(logging.DEBUG):
        logger.debug('Start')

    global IX_LAST_BENCHMARK

    measurement_data = [None, None, None, 0, 0]

    measurement_data[IX_LAST_BENCHMARK] = datetime.datetime.now()

    result_file = create_result_file(logger, config)

    measurement_data_result_file = (measurement_data, result_file)

    if logger.isEnabledFor(logging.DEBUG):
        logger.debug('End')

    return measurement_data_result_file


# ------------------------------------------------------------------------------
# Loading the bulk file into memory.
# ------------------------------------------------------------------------------

def get_bulk_data_partitions(logger, config):
    if logger.isEnabledFor(logging.DEBUG):
        logger.debug('Start')

    with open(os.path.abspath(config['file.bulk.name'])) as csv_file:
        bulk_data = [tuple(line) for line in csv.reader(csv_file, delimiter=config['file.bulk.delimiter'])]

    del bulk_data[0]

    bulk_data_partitions = [[] for _ in range(0, config['benchmark.number.partitions'])]

    # ------------------------------------------------------------------------------
    # Loading the bulk file into memory.
    # ------------------------------------------------------------------------------

    for key_data_tuple in bulk_data:
        key = key_data_tuple[0]
        partition_key = (ord(key[0]) * 256 + ord(key[1])) % config['benchmark.number.partitions']
        bulk_data_partition = bulk_data_partitions[partition_key]
        bulk_data_partition.append([key_data_tuple])
        bulk_data_partitions[partition_key] = bulk_data_partition

    logger.info('Start Distribution of the data in the partitions')

    for partition_key in range(0, config['benchmark.number.partitions']):
        logger.info('Partition p' + '{:0>5d}'.format(partition_key) + ' contains ' + '{0:n}'.format(len(bulk_data_partitions[partition_key])) + ' rows')

    logger.info('End   Distribution of the data in the partitions')

    if logger.isEnabledFor(logging.DEBUG):
        logger.debug('End')

    return bulk_data_partitions


# ------------------------------------------------------------------------------
# Loading the configuration parameters into memory.
# ------------------------------------------------------------------------------

def get_config(logger):
    if logger.isEnabledFor(logging.DEBUG):
        logger.debug('Start')

    global FILE_CONFIGURATION_NAME_PYTHON

    config_parser = configparser.ConfigParser()
    config_parser.read(FILE_CONFIGURATION_NAME_PYTHON)

    config = dict()

    config['benchmark.batch.size'] = int(config_parser['DEFAULT']['benchmark.batch.size'])
    config['benchmark.comment'] = config_parser['DEFAULT']['benchmark.comment']
    config['benchmark.core.multiplier'] = int(config_parser['DEFAULT']['benchmark.core.multiplier'])
    config['benchmark.database'] = config_parser['DEFAULT']['benchmark.database']
    config['benchmark.host.name'] = config_parser['DEFAULT']['benchmark.host.name']
    config['benchmark.id'] = config_parser['DEFAULT']['benchmark.id']
    config['benchmark.number.cores'] = int(config_parser['DEFAULT']['benchmark.number.cores'])
    config['benchmark.number.partitions'] = int(config_parser['DEFAULT']['benchmark.number.partitions'])
    config['benchmark.os'] = config_parser['DEFAULT']['benchmark.os']
    config['benchmark.release'] = config_parser['DEFAULT']['benchmark.release']
    config['benchmark.transaction.size'] = int(config_parser['DEFAULT']['benchmark.transaction.size'])
    config['benchmark.trials'] = int(config_parser['DEFAULT']['benchmark.trials'])
    config['benchmark.user.name'] = config_parser['DEFAULT']['benchmark.user.name']

    config['connection.fetch.size'] = int(config_parser['DEFAULT']['connection.fetch.size'])
    config['connection.host'] = config_parser['DEFAULT']['connection.host']
    config['connection.password'] = config_parser['DEFAULT']['connection.password']
    config['connection.port'] = int(config_parser['DEFAULT']['connection.port'])
    config['connection.service'] = config_parser['DEFAULT']['connection.service']
    config['connection.user'] = config_parser['DEFAULT']['connection.user']

    config['file.bulk.delimiter'] = str(config_parser['DEFAULT']['file.bulk.delimiter']).replace('TAB', '\t')
    config['file.bulk.length'] = int(config_parser['DEFAULT']['file.bulk.length'])
    config['file.bulk.name'] = config_parser['DEFAULT']['file.bulk.name']
    config['file.bulk.size'] = int(config_parser['DEFAULT']['file.bulk.size'])
    config['file.configuration.name.python'] = config_parser['DEFAULT']['file.configuration.name.python']
    config['file.result.delimiter'] = str(config_parser['DEFAULT']['file.result.delimiter']).replace('TAB', '\t')
    config['file.result.name'] = config_parser['DEFAULT']['file.result.name']

    config['sql.create'] = config_parser['DEFAULT']['sql.create']
    config['sql.drop'] = config_parser['DEFAULT']['sql.drop']
    config['sql.insert'] = config_parser['DEFAULT']['sql.insert'].replace(':key', ':1').replace(':data', ':2')
    config['sql.select'] = config_parser['DEFAULT']['sql.select']

    if logger.isEnabledFor(logging.DEBUG):
        logger.debug('End')

    return config


# ------------------------------------------------------------------------------
# Performing the insert operations.
# ------------------------------------------------------------------------------

def insert(logger, config, connection, cursor, bulk_data_partition):
    if logger.isEnabledFor(logging.DEBUG):
        logger.debug('Start')

    count = 0
    batch_data = list()

    for [key_data_tuple] in bulk_data_partition:
        count += 1

        if config['benchmark.batch.size'] == 0:
            cursor.execute(config['sql.insert'], [key_data_tuple[0], key_data_tuple[1]])
        else:
            batch_data.append(key_data_tuple)
            if count % config['benchmark.batch.size'] == 0:
                cursor.executemany(config['sql.insert'], batch_data)
                batch_data = list()

        if config['benchmark.transaction.size'] > 0 and count % config['benchmark.transaction.size'] == 0:
            connection.commit()

    if config['benchmark.batch.size'] > 0 and batch_data.__len__() > 0:
        cursor.executemany(config['sql.insert'], batch_data)

    if config['benchmark.transaction.size'] == 0 or count % config['benchmark.transaction.size'] != 0:
        connection.commit()

    if logger.isEnabledFor(logging.DEBUG):
        logger.debug('End')


# ------------------------------------------------------------------------------
# Main routine.
# ------------------------------------------------------------------------------

def main():
    with open('src_python/logging.yaml', 'r') as f:
        log_config = yaml.safe_load(f.read())

    logging.config.dictConfig(log_config)

    logger = logging.getLogger('OraBench.py')
    logger.setLevel(logging.DEBUG)

    if logger.isEnabledFor(logging.DEBUG):
        logger.debug('Start')

    logger.info('Start OraBench.py')

    locale.setlocale(locale.LC_ALL, 'de_DE.utf8')

    run_benchmark(logger)

    logger.info('End   OraBench.py')

    if logger.isEnabledFor(logging.DEBUG):
        logger.debug('End')


# ------------------------------------------------------------------------------
# Performing the benchmark run.
# ------------------------------------------------------------------------------

def run_benchmark(logger):
    if logger.isEnabledFor(logging.DEBUG):
        logger.debug('Start')

    config = get_config(logger)

    measurement_data_result_file = create_result_measuring_point_start_benchmark(logger, config)

    measurement_data = measurement_data_result_file[0]
    result_file = measurement_data_result_file[1]

    bulk_data_partitions = get_bulk_data_partitions(logger, config)

    connections_cursors = create_database_objects(logger, config)

    connections = connections_cursors[0]
    cursors = connections_cursors[1]

    for trial_number in range(0, config['benchmark.trials']):
        run_trial(logger, config, connections, cursors, bulk_data_partitions, measurement_data, result_file, trial_number + 1)

    for cursor in cursors:
        cursor.close()

    for connection in connections:
        connection.close()

    create_result_measuring_point_end(logger, config, result_file, measurement_data, 'benchmark')

    if logger.isEnabledFor(logging.DEBUG):
        logger.debug('End')


# ------------------------------------------------------------------------------
# Performing the insert operations.
# ------------------------------------------------------------------------------

def run_insert(logger, config, connections, cursors, bulk_data_partitions, result_file, measurement_data, trial_number):
    if logger.isEnabledFor(logging.DEBUG):
        logger.debug('Start')

    measurement_data = create_result_measuring_point_start(logger, measurement_data, 'query')

    threads = list()

    for partition_key in range(0, config['benchmark.number.partitions']):
        if config['benchmark.core.multiplier'] == 0:
            insert(logger, config, connections[partition_key], cursors[partition_key], bulk_data_partitions[partition_key])
        else:
            thread = threading.Thread(target=insert, args=(config, connections[partition_key], cursors[partition_key], bulk_data_partitions[partition_key],))
            threads.append(thread)
            thread.start()

    if config['benchmark.core.multiplier'] > 0:
        for thread in threads:
            thread.join()

    create_result_measuring_point_end(logger, config, result_file, measurement_data, 'query', trial_number, config['sql.insert'], 'insert')

    if logger.isEnabledFor(logging.DEBUG):
        logger.debug('End')

    return measurement_data


# ------------------------------------------------------------------------------
# Performing the select operations.
# ------------------------------------------------------------------------------

def run_select(logger, config, cursors, bulk_data_partitions, result_file, measurement_data, trial_number):
    if logger.isEnabledFor(logging.DEBUG):
        logger.debug('Start')

    measurement_data = create_result_measuring_point_start(logger, measurement_data, 'query')

    threads = list()

    for partition_key in range(0, config['benchmark.number.partitions']):
        if config['benchmark.core.multiplier'] == 0:
            select(logger, cursors[partition_key], bulk_data_partitions[partition_key], partition_key, config['sql.select'])
        else:
            thread = threading.Thread(target=select, args=(cursors[partition_key], bulk_data_partitions[partition_key], partition_key, config['sql.select'],))
            threads.append(thread)
            thread.start()

    if config['benchmark.core.multiplier'] > 0:
        for thread in threads:
            thread.join()

    create_result_measuring_point_end(logger, config, result_file, measurement_data, 'query', trial_number, config['sql.select'], 'select')

    if logger.isEnabledFor(logging.DEBUG):
        logger.debug('End')

    return measurement_data


# ------------------------------------------------------------------------------
# Performing one trial.
# ------------------------------------------------------------------------------

def run_trial(logger, config, connections, cursors, bulk_data_partitions, measurement_data, result_file, trial_number):
    if logger.isEnabledFor(logging.DEBUG):
        logger.debug('Start')

    measurement_data = create_result_measuring_point_start(logger, measurement_data, 'trial')

    logger.info('Start trial no. ' + str(trial_number))

    try:
        cursors[0].execute(config['sql.create'])
        logger.debug('last DDL statement=' + config['sql.create'])
    except cx_Oracle.DatabaseError:
        cursors[0].execute(config['sql.drop'])
        cursors[0].execute(config['sql.create'])
        logger.debug('last DDL statement after DROP=' + config['sql.create'])

    run_insert(logger, config, connections, cursors, bulk_data_partitions, result_file, measurement_data, trial_number)

    run_select(logger, config, cursors, bulk_data_partitions, result_file, measurement_data, trial_number)

    cursors[0].execute(config['sql.drop'])
    logger.debug('last DDL statement=' + config['sql.drop'])

    create_result_measuring_point_end(logger, config, result_file, measurement_data, 'trial', trial_number)

    if logger.isEnabledFor(logging.DEBUG):
        logger.debug('End')

    return measurement_data


# ------------------------------------------------------------------------------
# Performing the select operations.
# ------------------------------------------------------------------------------

def select(logger, cursor, bulk_size_partition, partition_key, sql_statement):
    if logger.isEnabledFor(logging.DEBUG):
        logger.debug('Start')

    count = 0

    cursor.execute(sql_statement + ' where partition_key = ' + str(partition_key))

    for _ in cursor:
        count += 1

    if count != len(bulk_size_partition):
        logger.error('Number rows: expected=' + str(len(bulk_size_partition)) + ' - found=' + str(count))
        sys.exit(1)

    if logger.isEnabledFor(logging.DEBUG):
        logger.debug('End')


# ------------------------------------------------------------------------------
# Program start.
# ------------------------------------------------------------------------------

main()
